File: Assets/Scripts/Sub_Movement.js
#pragma strict

var movementSpeed : float = 1000;
var surfaceHeight : float = 448.2;

private var atSurface : boolean = false;


function Update ()
{
	/***************
	 Movement
	***************/
	
	if(Input.GetKey("w"))
	{
		if(transform.position.y < surfaceHeight)
			rigidbody.AddForce(Vector3.up * movementSpeed, ForceMode.Impulse); 
			//rigidbody.velocity = Vector3(0,1,0)*movementSpeed;
	}
	if(Input.GetKey("s"))
	{
		rigidbody.AddForce(Vector3.down * movementSpeed, ForceMode.Impulse); 
		//rigidbody.velocity = Vector3(0,-1,0)*movementSpeed;
	}
	if(Input.GetKey("a"))
	{
		rigidbody.AddForce(Vector3.left * movementSpeed, ForceMode.Impulse);
		//rigidbody.velocity = Vector3(-1,0,0)*movementSpeed;
	}
	if(Input.GetKey("d"))
	{
		rigidbody.AddForce(Vector3.right * movementSpeed, ForceMode.Impulse);
		//rigidbody.velocity = Vector3(1,0,0)*movementSpeed;
	}
	if(Input.GetKey("up"))
	{
		rigidbody.AddForce(Vector3.forward * movementSpeed, ForceMode.Impulse);
		//rigidbody.velocity = Vector3(0,0,1)*movementSpeed;
	}
	if(Input.GetKey("down"))
	{
		rigidbody.AddForce(Vector3.back * movementSpeed, ForceMode.Impulse);
		//rigidbody.velocity = Vector3(0,0,-1)*movementSpeed;
	}
	if(Input.GetKey("space"))
	{
		//rigidbody.AddForce(Vector3.back * movementSpeed);
		rigidbody.velocity = Vector3(0,0,0);
	}		
	
	//Debug.Log("Velocity: (" + rigidbody.velocity.x + ", " + rigidbody.velocity.y + ", " + rigidbody.velocity.z + ").");
	
	/***************
	 World Boundaries
	***************/
	var currentX = transform.position.x;
	var currentY = transform.position.y;
	var currentZ = transform.position.z;
	
	if(currentX >= 1000)
		transform.position = Vector3(1000, currentY, currentZ);
	if(currentX <= -1000)
		transform.position = Vector3(-1000, currentY, currentZ);
		
	if(currentZ >= 1000)
		transform.position = Vector3(currentX, currentY, 1000);
	if(currentZ <= -1000)
		transform.position = Vector3(currentX, currentY, -1000);
		
	//if(currentY >= surfaceHeight)
	//	transform.position = Vector3(currentX, surfaceHeight-.1, currentZ);
	
}

File: Assets/Scripts/SubMovement.cs
using UnityEngine;
using System.Collections;

public class SubMovement : MonoBehaviour
{
	// public variables

	public float movementSpeed; // = 100f;
	public float rotationSpeed; // = 5f;
	public float maxSpeed;

	public GameObject[] propellers;
	public float propChangeRate;
	public ParticleSystem[] bubbles;

	public AudioClip hit;


	// private variables
	private float surfaceHeight = 450.2f;


	void Start()
	{
		rigidbody.freezeRotation = true;
	}
	

	void FixedUpdate()
	{
	
		if(transform.position.y < surfaceHeight)
		{
			if(Input.GetKey("w"))
			{
				up();
			}
		}

		if(Input.GetKey("s"))
		{
			down();
		}

		if(Input.GetKey("up"))
		{
			forward();
			foreach(GameObject prop in propellers)
			{	
				prop.animation.CrossFade("FastSpin", propChangeRate);
			}

			foreach(ParticleSystem bubble in bubbles)
			{
				bubble.emissionRate = 100f;
			}
		}

		else
		{
			foreach(GameObject prop in propellers)
			{	
				prop.animation.CrossFade("SlowSpin", propChangeRate);
			}

			foreach(ParticleSystem bubble in bubbles)
			{
				bubble.emissionRate = 10f;
			}
		}

		if(Input.GetKey("down"))
		{
			backward();
		}
		
		// Rotation
		float rotation = Input.GetAxis("Horizontal") * rotationSpeed * Time.deltaTime;
		rotate(rotation);


		// Keep in water
		if(transform.position.y >= surfaceHeight)
		{
			Vector3 temp = new Vector3(rigidbody.velocity.x, 0, rigidbody.velocity.z);
			rigidbody.velocity = temp;
		}
		
	}
	

	/***************************
	 Control Functions
	**************************/
	void up()
	{
		rigidbody.AddForce(Vector3.up * movementSpeed, ForceMode.Acceleration); 
	}

	void down()
	{
		rigidbody.AddForce(Vector3.down * movementSpeed, ForceMode.Acceleration); 	
	}

	void forward()
	{
		if(rigidbody.velocity.magnitude < maxSpeed)
			rigidbody.AddForce(transform.forward * movementSpeed, ForceMode.Acceleration);

	}

	void backward()
	{
		rigidbody.AddForce(-1 * transform.forward * movementSpeed, ForceMode.Acceleration);	
	}

	void rotate(float rotation)
	{
		rigidbody.freezeRotation = false;
		transform.Rotate(0, rotation, 0);
		rigidbody.freezeRotation = true;
	}

	void OnCollisionEnter(Collision collision)
	{
		audio.PlayOneShot(hit);
	}
	
}


